File: src/agents/portfolio_manager.py
import json
from langchain_core.messages import HumanMessage, SystemMessage
from langchain_core.prompts import ChatPromptTemplate

from graph.state import AgentState, show_agent_reasoning
from pydantic import BaseModel, Field
from typing_extensions import Literal
from utils.progress import progress
from utils.llm import call_llm
from datetime import datetime
from llm.models import get_model, ModelProvider


class PortfolioDecision(BaseModel):
    action: Literal["buy", "sell", "short", "cover", "hold"]
    quantity: int = Field(description="Number of shares to trade")
    confidence: float = Field(description="Confidence in the decision, between 0.0 and 100.0")
    reasoning: str = Field(description="Reasoning for the decision")


class PortfolioManagerOutput(BaseModel):
    decisions: dict[str, PortfolioDecision] = Field(description="Dictionary of ticker to trading decisions")


def parse_llm_response(response_text):
    """Parse JSON from LLM response"""
    try:
        # Try to extract JSON content if it's wrapped in markdown code blocks
        if "```json" in response_text:
            start_idx = response_text.find("```json") + 7
            end_idx = response_text.find("```", start_idx)
            json_str = response_text[start_idx:end_idx].strip()
            return json.loads(json_str)
        # Try to parse the entire response as JSON
        else:
            return json.loads(response_text)
    except Exception as e:
        print(f"Error parsing LLM response: {e}")
        return {"error": "Could not parse response"}


##### Portfolio Management Agent #####
def portfolio_management_agent(state):
    """Portfolio Management Agent"""
    # In LangGraph 0.2.56, state might be a dict instead of AgentState object
    # Handle both cases for compatibility
    if isinstance(state, dict):
        data = state["data"]
        metadata = state["metadata"]
        messages = state["messages"]
    else:
        data = state.data
        metadata = state.metadata
        messages = state.messages
    
    tickers = data["tickers"]
    portfolio = data["portfolio"]
    signals = data["analyst_signals"]
    model_name = metadata.get("model_name", "gpt-4o")
    model_provider = metadata.get("model_provider", "OpenAI")
    
    # Get trade history and frequency from state 
    trade_history = data.get("trade_history", {})
    trading_frequency = data.get("trading_frequency", {})
    current_date = data.get("current_date", datetime.now().strftime("%Y-%m-%d"))
    
    # Generate prompt including trade history
    prompt = f"""You are the Portfolio Management Agent of an AI-powered hedge fund.

Current Date: {current_date}

Your role is to make the final trading decisions based on analyst recommendations and risk management guidelines.

Current Portfolio:
{json.dumps(portfolio, indent=2)}

Risk Analysis:
{json.dumps(signals.get("risk_management_agent", {}), indent=2)}

Current Analyst Signals:
{json.dumps({k: v for k, v in signals.items() if k != "risk_management_agent"}, indent=2)}

Trading Frequency Metrics:
{json.dumps(trading_frequency, indent=2)}

## Overtrading Protection Rules
1. Maximum 5 trades per day across all tickers
2. Wait at least 2 days after a trade before trading same ticker again, unless:
   - New significant information justifies immediate action
   - A clear reversal signal is present with high conviction
3. Maximum 3 trades per week per ticker
4. Today's trade count so far: {trading_frequency.get('today_trade_count', 0)}
5. Never make multiple trades in the same ticker on the same day

For each ticker, evaluate all signals and determine the optimal action:
1. "buy": Buy shares if bullish signals are strong, we have available cash, and haven't traded the ticker recently
2. "sell": Sell shares if bearish signals are strong and we own shares
3. "short": Short stock if very bearish signals exist AND we have margin capability
4. "cover": Cover short position if signals turn neutral or bullish
5. "hold": Take no action (default)

For each action, provide a quantity that:
- Respects the position limits from risk management
- Accounts for available cash
- Is proportional to the conviction level
- Considers existing positions
- Prevents overtrading/excessive turnover

Respond with JSON format with decisions for each ticker:
{{
  "TICKER1": {{
    "action": "buy"|"sell"|"short"|"cover"|"hold",
    "quantity": 10
  }},
  "TICKER2": {{ ... }}
}}
"""

    # Get reasoning and decisions from LLM
    llm = get_model(model_name, ModelProvider(model_provider))
    messages_for_llm = [
        HumanMessage(content=prompt)
    ]
    try:
        response = llm.invoke(messages_for_llm)
        reasoning = response.content
        
        # Check if we're approaching trading limits
        today_count = trading_frequency.get('today_trade_count', 0)
        max_daily_trades = 5  # Maximum allowed trades per day
        approaching_limit = today_count >= max_daily_trades - 2  # Warning if we have 3+ trades already
        
        # Parse the output as-is
        parsed_response = parse_llm_response(reasoning)
        
        # Apply overtrading protection
        if approaching_limit:
            print(f"{Fore.YELLOW}WARNING: Approaching daily trade limit ({today_count}/{max_daily_trades}){Style.RESET_ALL}")
            
            # If we're at the limit, override decisions to "hold"
            if today_count >= max_daily_trades:
                print(f"{Fore.RED}ALERT: Daily trade limit reached. Forcing HOLD on all positions.{Style.RESET_ALL}")
                for ticker in parsed_response:
                    if parsed_response[ticker]["action"] != "hold":
                        parsed_response[ticker]["action"] = "hold"
                        parsed_response[ticker]["quantity"] = 0
                        parsed_response[ticker]["override_reason"] = "Daily trade limit reached"
        
        # Check for recent trades on each ticker
        for ticker in list(parsed_response.keys()):
            ticker_history = trade_history.get(ticker, [])
            if ticker_history and isinstance(ticker_history, list) and len(ticker_history) > 0:
                first_trade = ticker_history[0]
                days_ago = first_trade.get("days_ago")
                
                if days_ago is not None and days_ago < 2:
                    # We traded this ticker in the last 2 days
                    print(f"{Fore.YELLOW}Recent trade detected for {ticker} ({days_ago} days ago){Style.RESET_ALL}")
                    
                    # Check if there's a strong consensus from multiple analysts
                    ticker_signals_count = sum(1 for k, v in signals.items() 
                                              if k != "risk_management_agent" and ticker in v)
                    strong_consensus = ticker_signals_count >= 3
                    
                    decision = parsed_response[ticker]
                    if not strong_consensus and decision["action"] != "hold":
                        decision["original_action"] = decision["action"]
                        decision["action"] = "hold"
                        decision["quantity"] = 0
                        decision["override_reason"] = f"Recent trade ({days_ago} days ago) without strong new consensus"
        
        # Convert parsed_response to string for JSON response
        final_decisions = {}
        for ticker, decision in parsed_response.items():
            final_decisions[ticker] = {
                "action": decision.get("action", "hold"),
                "quantity": decision.get("quantity", 0),
                "confidence": decision.get("confidence", 50.0)
            }
            
    except Exception as e:
        print(f"Error in portfolio management agent: {e}")
        final_decisions = {}
        reasoning = f"Error: {str(e)}"

    # If show_reasoning is enabled, print it
    if metadata.get("show_reasoning", False):
        print(f"\n{Fore.BLUE}Portfolio Management Agent Reasoning:{Style.RESET_ALL}")
        print(reasoning)

    # Return the response as a system message
    if isinstance(state, dict):
        state["messages"].append(SystemMessage(content=json.dumps(final_decisions)))
    else:
        state.messages.append(SystemMessage(content=json.dumps(final_decisions)))
        
    return state


def generate_trading_decision(
    tickers: list[str],
    signals_by_ticker: dict[str, dict],
    current_prices: dict[str, float],
    max_shares: dict[str, int],
    portfolio: dict[str, float],
    model_name: str,
    model_provider: str,
) -> PortfolioManagerOutput:
    """Attempts to get a decision from the LLM with retry logic"""
    # Create the prompt template
    template = ChatPromptTemplate.from_messages(
        [
            (
              "system",
              """You are a portfolio manager making final trading decisions based on multiple tickers.

              Trading Rules:
              - For long positions:
                * Only buy if you have available cash
                * Only sell if you currently hold long shares of that ticker
                * Sell quantity must be ≤ current long position shares
                * Buy quantity must be ≤ max_shares for that ticker
              
              - For short positions:
                * Only short if you have available margin (50% of position value required)
                * Only cover if you currently have short shares of that ticker
                * Cover quantity must be ≤ current short position shares
                * Short quantity must respect margin requirements
              
              - The max_shares values are pre-calculated to respect position limits
              - Consider both long and short opportunities based on signals
              - Maintain appropriate risk management with both long and short exposure

              Available Actions:
              - "buy": Open or add to long position
              - "sell": Close or reduce long position
              - "short": Open or add to short position
              - "cover": Close or reduce short position
              - "hold": No action

              Inputs:
              - signals_by_ticker: dictionary of ticker → signals
              - max_shares: maximum shares allowed per ticker
              - portfolio_cash: current cash in portfolio
              - portfolio_positions: current positions (both long and short)
              - current_prices: current prices for each ticker
              - margin_requirement: current margin requirement for short positions
              """,
            ),
            (
              "human",
              """Based on the team's analysis, make your trading decisions for each ticker.

              Here are the signals by ticker:
              {signals_by_ticker}

              Current Prices:
              {current_prices}

              Maximum Shares Allowed For Purchases:
              {max_shares}

              Portfolio Cash: {portfolio_cash}
              Current Positions: {portfolio_positions}
              Current Margin Requirement: {margin_requirement}

              Output strictly in JSON with the following structure:
              {{
                "decisions": {{
                  "TICKER1": {{
                    "action": "buy/sell/short/cover/hold",
                    "quantity": integer,
                    "confidence": float between 0 and 100,
                    "reasoning": "string"
                  }},
                  "TICKER2": {{
                    ...
                  }},
                  ...
                }}
              }}
              """,
            ),
        ]
    )

    # Generate the prompt
    prompt = template.invoke(
        {
            "signals_by_ticker": json.dumps(signals_by_ticker, indent=2),
            "current_prices": json.dumps(current_prices, indent=2),
            "max_shares": json.dumps(max_shares, indent=2),
            "portfolio_cash": f"{portfolio.get('cash', 0):.2f}",
            "portfolio_positions": json.dumps(portfolio.get('positions', {}), indent=2),
            "margin_requirement": f"{portfolio.get('margin_requirement', 0):.2f}",
        }
    )

    # Create default factory for PortfolioManagerOutput
    def create_default_portfolio_output():
        return PortfolioManagerOutput(decisions={ticker: PortfolioDecision(action="hold", quantity=0, confidence=0.0, reasoning="Error in portfolio management, defaulting to hold") for ticker in tickers})

    return call_llm(prompt=prompt, model_name=model_name, model_provider=model_provider, pydantic_model=PortfolioManagerOutput, agent_name="portfolio_management_agent", default_factory=create_default_portfolio_output)

File: src/agents/portfolio_manager.rs
// Source: src/agents/portfolio_manager.py
//! Sibling to src/agents/portfolio_manager.py
//! Makes the final trading decisions, enforcing position sizing limits and capital/margin constraints.

use anyhow::{Result, Context};
use std::collections::HashMap;
use crate::graph::state::AgentState;

/// Struct mirroring the Pydantic `PortfolioDecision` model.
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
pub struct PortfolioDecision {
    pub action: String,      // "buy" | "sell" | "short" | "cover" | "hold"
    pub quantity: u32,
    pub confidence: u32,     // 0-100
    pub reasoning: String,
}

/// Struct mirroring the Pydantic `PortfolioManagerOutput` model.
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
pub struct PortfolioManagerOutput {
    pub decisions: HashMap<String, PortfolioDecision>,
}

/// Main entry node for the portfolio management agent.
pub async fn portfolio_management_agent(state: &mut AgentState, agent_id: &str) -> Result<()> {
    println!("Running Portfolio Management Agent: {}", agent_id);

    let tickers_json = state.data.get("tickers")
        .context("Missing tickers in state data")?;
    let tickers: Vec<String> = serde_json::from_value(tickers_json.clone())?;

    let portfolio = state.data.get("portfolio")
        .context("Missing portfolio in portfolio_manager")?;
    
    let analyst_signals = state.data.get("analyst_signals")
        .context("Missing analyst_signals in state data")?;

    let mut current_prices = HashMap::new();
    let mut max_shares = HashMap::new();
    let mut signals_by_ticker = HashMap::new();

    // Risk manager ID mapping
    let risk_manager_id = "risk_management_agent";
    let risk_data = analyst_signals.get(risk_manager_id);

    for ticker in &tickers {
        let remaining_limit = risk_data
            .and_then(|r| r.get(ticker))
            .and_then(|t| t.get("remaining_position_limit"))
            .and_then(|l| l.as_f64())
            .unwrap_or(0.0);

        let current_price = risk_data
            .and_then(|r| r.get(ticker))
            .and_then(|t| t.get("current_price"))
            .and_then(|p| p.as_f64())
            .unwrap_or(0.0);

        current_prices.insert(ticker.clone(), current_price);

        let max_qty = if current_price > 0.0 {
            (remaining_limit / current_price) as u32
        } else {
            0
        };
        max_shares.insert(ticker.clone(), max_qty);

        // Gather signals from other analysts (excluding risk manager)
        let mut ticker_signals = HashMap::new();
        if let Some(signals_obj) = analyst_signals.as_object() {
            for (agent, sigs) in signals_obj {
                if agent != risk_manager_id {
                    if let Some(t_sig) = sigs.get(ticker) {
                        let sig = t_sig.get("signal").and_then(|s| s.as_str()).unwrap_or("neutral").to_string();
                        let conf = t_sig.get("confidence").and_then(|c| c.as_u64()).unwrap_or(50) as u32;
                        ticker_signals.insert(agent.clone(), (sig, conf));
                    }
                }
            }
        }
        signals_by_ticker.insert(ticker.clone(), ticker_signals);
    }

    // Generate Decisions
    let result = generate_trading_decision(
        tickers,
        signals_by_ticker,
        current_prices,
        max_shares,
        portfolio,
    )?;

    // Append final message to state messages
    let decision_value = serde_json::to_value(result.clone())?;
    state.messages.push(decision_value);

    // Save final decision in state data
    state.data.insert("decisions".to_string(), serde_json::to_value(result.decisions)?);

    Ok(())
}

pub fn compute_allowed_actions(
    tickers: &[String],
    current_prices: &HashMap<String, f64>,
    max_shares: &HashMap<String, u32>,
    portfolio: &serde_json::Value,
) -> HashMap<String, HashMap<String, u32>> {
    let mut allowed = HashMap::new();
    let cash = portfolio.get("cash").and_then(|v| v.as_f64()).unwrap_or(100000.0);
    let positions_map = portfolio.get("positions").and_then(|p| p.as_object());

    for ticker in tickers {
        let price = current_prices.get(ticker).copied().unwrap_or(0.0);
        let max_qty = max_shares.get(ticker).copied().unwrap_or(0);

        let long_shares = positions_map
            .and_then(|p| p.get(ticker))
            .and_then(|pos| pos.get("long"))
            .and_then(|l| l.as_i64())
            .unwrap_or(0) as u32;

        let short_shares = positions_map
            .and_then(|p| p.get(ticker))
            .and_then(|pos| pos.get("short"))
            .and_then(|s| s.as_i64())
            .unwrap_or(0) as u32;

        let mut actions = HashMap::new();
        actions.insert("hold".to_string(), 0);

        // Long capacity
        if long_shares > 0 {
            actions.insert("sell".to_string(), long_shares);
        }
        if cash > 0.0 && price > 0.0 {
            let max_buy_cash = (cash / price) as u32;
            let max_buy = std::cmp::min(max_qty, max_buy_cash);
            if max_buy > 0 {
                actions.insert("buy".to_string(), max_buy);
            }
        }

        // Short capacity
        if short_shares > 0 {
            actions.insert("cover".to_string(), short_shares);
        }
        if price > 0.0 && max_qty > 0 {
            let max_short = max_qty; // Sized by risk limit and cash
            if max_short > 0 {
                actions.insert("short".to_string(), max_short);
            }
        }

        allowed.insert(ticker.clone(), actions);
    }
    allowed
}

pub fn generate_trading_decision(
    tickers: Vec<String>,
    signals_by_ticker: HashMap<String, HashMap<String, (String, u32)>>,
    current_prices: HashMap<String, f64>,
    max_shares: HashMap<String, u32>,
    portfolio: &serde_json::Value,
) -> Result<PortfolioManagerOutput> {
    let allowed_actions = compute_allowed_actions(&tickers, &current_prices, &max_shares, portfolio);
    let mut decisions = HashMap::new();

    for ticker in tickers {
        let allowed = allowed_actions.get(&ticker).context("Missing allowed actions for ticker")?;
        let ticker_signals = signals_by_ticker.get(&ticker).context("Missing signals for ticker")?;

        // Consensus signal voting
        let mut bull_weight = 0.0;
        let mut bear_weight = 0.0;
        for (_agent, (sig, conf)) in ticker_signals {
            let weight = *conf as f64 / 100.0;
            if sig == "bullish" {
                bull_weight += weight;
            } else if sig == "bearish" {
                bear_weight += weight;
            }
        }

        let mut action = "hold".to_string();
        let mut quantity = 0;
        let mut reasoning = "Neutral stance".to_string();

        if bull_weight > bear_weight {
            // We want to buy or cover
            if allowed.contains_key("cover") {
                action = "cover".to_string();
                quantity = *allowed.get("cover").unwrap();
                reasoning = "Covering short position based on bullish consensus".to_string();
            } else if allowed.contains_key("buy") {
                action = "buy".to_string();
                quantity = *allowed.get("buy").unwrap();
                reasoning = "Initiating long buy based on bullish consensus".to_string();
            }
        } else if bear_weight > bull_weight {
            // We want to sell or short
            if allowed.contains_key("sell") {
                action = "sell".to_string();
                quantity = *allowed.get("sell").unwrap();
                reasoning = "Selling long position based on bearish consensus".to_string();
            } else if allowed.contains_key("short") {
                action = "short".to_string();
                quantity = *allowed.get("short").unwrap();
                reasoning = "Initiating short position based on bearish consensus".to_string();
            }
        }

        let conf = if bull_weight + bear_weight > 0.0 {
            (bull_weight.max(bear_weight) / (bull_weight + bear_weight) * 100.0) as u32
        } else {
            100
        };

        decisions.insert(
            ticker,
            PortfolioDecision {
                action,
                quantity,
                confidence: conf,
                reasoning,
            },
        );
    }

    Ok(PortfolioManagerOutput { decisions })
}
